### ISSUE TITLE
Rule S4822: Promise rejections should not be caught by 'try' block

### ISSUE THREAD

**Issue by maximelkin** [SonarSource/SonarTS#716]
**I want to request a feature.**
Rule to detect this cases 
```ts
try { 
   return promise;
} catch (e) {}
```
which is possible error, because promise error will not be catch.

**Comment by maximelkin** [comment]
Can i implement this rule?

**Comment by vilchik-elena** [comment]
Rule should raise an issue when the only statement of try is `return <promise>` or `<promise>.then()` or `<promise>.catch`
```ts
try {
  return promise;
} catch {}

try {
  promise.then()
} catch {}
```

@maximelkin Yeah, go for it! 
Are you sure that suggesting `await` is always valid? In some cases `async`/`await` might not be available

**Comment by maximelkin** [comment]
Also i think what this statement should not be only one so
```
try {
const a = 12;
return fetch('../...');
} catch (e) {
}
``` 
Should be warned to

**Comment by vilchik-elena** [comment]
@maximelkin 
I've created RSPEC: https://jira.sonarsource.com/browse/RSPEC-4822

I think we should detect following pattern
1. find 'try'
2. all statements inside 'try' are
   * expression statement with expression promise type
   * return statement returning expression with promise type

**Comment by maximelkin** [comment]
No, I try to describe rule which help to catch forgotten await in try block, because this is bug or place which can cause misunderstand of code. For example:
```ts
// probably bug
async function apiCall(path: string) {
    try {
        return request(path);
    } catch (e) {
        if (e.statusCode === 429) {
            await delay(3000);
            return apiCall(path);
        }
    }
}
```

**Comment by vilchik-elena** [comment]
@maximelkin "No" what?
 The rule the way I formalized it will report for your code sample

**Comment by vilchik-elena** [comment]
@maximelkin concerning your comment
 
> Also i think what return statement should not be only one in block so
> ```ts
> try {
>     const a = 12;
>     return fetch('../...');
> } catch (e) {
> }
> ``` 
> Should be warned too


I agree it should be warned, but how to formalize the logic? we should be sure that `catch` is not for that additional statement. Indeed we know that `const a = 12` can't throw, but what if it's `const a = foo(12)` instead?

**Comment by maximelkin** [comment]
More complicated example: 
```ts
// probably bug
async function apiCall(path: string) {
    try {
        if (a === 12) {
            return request.get(path);
        } else {
            return request.post(path);
        }
    } catch (e) {
        if (e.statusCode === 429) {
            await delay(3000);
            return apiCall(path);
        }
    }
}
```
current realization in related pr detect this

**Comment by maximelkin** [comment]
So this rule should detect any return <value> inside try block, where value is promise
And should detect in any level of scopes, but only if return related to same function where try block is placed

**Comment by maximelkin** [comment]
Will be RSPEC changed?
1) rule should detect return inside try block on any level of blocks and not only as one line
2) rule should not detect floating promises - because this work for no-floating-promises rule (second code example)

**Comment by vilchik-elena** [comment]
1. 
Before we continue, let's make sure that we are on the same page. This rules is supposed to detect cases when developer mistakenly thinks that exception thrown by his promise will be caught by nesting `try`. 

Then comes the question: how we can know that `catch` is there for promise and not for something else inside `try` block.

I suggested that we check `try` blocks with only "promise" statements.

I feel like you don't like this approach because you want to raise issues for more sophisticated cases (e.g. promise is returned conditionally). But then we will have plenty FPs because developer `catch`es not promise, but something around it. Like here:
```
 try {
  const response = JSON.parse(err.responseText);
  return TPromise.wrapError(response.message);
 } catch (e) {
  // noop
 }
```

**Comment by vilchik-elena** [comment]
2. Our guidelines tell us, that we should not excluded some cases from one rule, because we expect another rule to raise an issue for the same code. We never know which rules will be activated and which not. And these 2 rules while both relate to promises, are detecting different problems.

**Comment by maximelkin** [comment]
Ok, I agreed with floating promises detection
About try with only promises - i think this make rule unpredictable
I want take some time to rethink rule design

**Comment by maximelkin** [comment]
Maybe use this rule only if async/await style in project used? (and special handling of always error-resolved methods or just Promise.reject)

**Comment by vilchik-elena** [comment]
> About try with only promises - i think this make rule unpredictable
what do you mean?

> Maybe use this rule only if async/await style in project used? (and special handling of always error-resolved methods or just Promise.reject)
Could you elaborate more? (it would be nice with examples)

**Comment by andrea-guarino-sonarsource** [comment]
#728 was closed due to inactivity. 
We will reconsider and reopen a PR related to this issue when we'll work on SonarTS 1.9
